### PR TITLE
Adds initial implementation of customer details page redesign

### PR DESCRIPTION
**Pull request by rhonsby**
### Description

This commit consists of the initial implementation of the customer details page redesign. 

Note that I hid the redesign behind an env variable because it isn't complete in its implementation. I'm doing this for the purpose of ensuring that these commits are smaller and more reviewable. There's still a few missing features:
* Ability to edit basic properties like name, email, phone number, etc.
* Ability to delete customer
* Ability to export data
* Ability to view notes

#### Before
image

#### After

image

#### After (non-live customer and unattached to company)
image

## Checklist

- [x] Everything passes when running `mix test`
- [x] Ran `mix format`
- [x] No frontend compilation warnings

### CODE REVIEW

File: assets/src/api.ts [hunk 1]
```diff
@@ -1269,7 +1270,7 @@ type BrowserSessionFilters = {
 export const fetchBrowserSessions = async (
   {customerId, isActive, limit = 100, sessionIds = []}: BrowserSessionFilters,
   token = getAccessToken()
-) => {
+): Promise<Array<BrowserSession>> => {
```
Comment (rhonsby) [inline assets/src/api.ts:1273]: We should ideally type all of our API responses, right?

File: assets/src/components/Dashboard.tsx [hunk 2]
```diff
@@ -421,7 +423,14 @@ const Dashboard = (props: RouteComponentProps) => {
             component={GettingStartedOverview}
           />
           <Route path="/account*" component={AccountOverview} />
-          <Route path="/customers/:id" component={CustomerDetailsPage} />
+          <Route
+            path="/customers/:id"
+            component={
+              isCustomerDetailsV2Enabled
+                ? CustomerDetailsPageV2
+                : CustomerDetailsPage
```
Comment (rhonsby) [inline assets/src/components/Dashboard.tsx:431]: As mentioned in the description, I'm hiding the redesign behind an env variable while I work on implementing the rest of the features. I didn't want to have these changes sitting on a branch for too long.

Comment (reichert621) [reply assets/src/components/Dashboard.tsx:431]: perfect! was going to recommend this approach anyway :+1:

Comment (reichert621) [reply assets/src/components/Dashboard.tsx:431]: we could also just create a new temporary route for v2, like `/v2/customers/:id` -- might make it easier to view both 🤷

File: assets/src/components/customers/CustomerUtils.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import * as API from '../../api';
+import logger from '../../logger';
+import {download} from '../../utils';
+
+export const exportCustomerData = async (customerId: string) => {
```
Comment (rhonsby) [inline assets/src/components/customers/CustomerUtils.ts:5]: Seemed like a fairly component-agnostic utility function so I extracted it into its own utils file.

File: lib/chat_api_web/views/customer_view.ex [hunk 4]
```diff
@@ -50,7 +50,8 @@ defmodule ChatApiWeb.CustomerView do
       current_url: customer.current_url,
       browser: customer.browser,
       os: customer.os,
-      metadata: customer.metadata
+      metadata: customer.metadata,
+      title: customer.name || customer.email || "Anonymous User"
```
Comment (rhonsby) [inline lib/chat_api_web/views/customer_view.ex:54]: I noticed we had similar logic in different places, but would render different content. I figured this would be a good way to consolidate that so we don't have to duplicate that logic in too many places.

File: assets/src/components/Dashboard.tsx (hunk 2 above)
Comment (rhonsby) [reply assets/src/components/Dashboard.tsx:431]: Yeah, that sounds like a good idea! I'll add that. It'll be nice to be able to view it in prod.

File: assets/src/components/customers/CustomerDetailsSidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+      <Box mb={2}>
+        <Text strong>{title}</Text>
+      </Box>
+      {children}
+    </Box>
+  );
+};
+
+const CustomerDetailsProperty = ({
+  Icon,
```
Comment (reichert621) [inline assets/src/components/customers/CustomerDetailsSidebar.tsx:242]: nit -- i think react props should always be lowercase? 🤔 

but i see this is passing in the icon class/function rather than the component itself, so i see why it's capitalized... i think i'd rather pass in the component itself even if we need to explicitly set the color every time, e.g.
```jsx
<CustomerDetailsProperty
  icon={<LinkOutlined style={{color: colors.primary}} />}
  name="Website"
  value={websiteUrl}
/>
```

File: assets/src/components/customers/CustomerDetailsSidebar.tsx [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+  name,
+  value,
+}: {
+  Icon: React.ComponentType<any>;
+  name: string;
+  value: any;
+}) => {
+  let valueComponent;
```
Comment (reichert621) [inline assets/src/components/customers/CustomerDetailsSidebar.tsx:250]: hmm i wonder if similar to the `icon` prop we should just enforce that this is a `ReactElement` type and explicitly pass in `<Text>{value || 'Unknown'}</Text>` when we need to 🤔 i kinda prefer the explicit nature of it, and in some cases we might want the fallback value to be something other than "unknown" 

(this behavior might also be confusing if a value of `0` is passed in, for example... though i suppose this edge case could be fixed by using `value ?? 'Unknown'` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)

otherwise, small nitpick but we can probably just assign the `valueComponent` in a ternary expression, e.g.
```js
const valueComponent = React.isValidElement(value) ? value : <Text>{value || 'Unknown'}</Text>;
```

File: assets/src/components/customers/CustomerDetailsSidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+      {valueComponent}
+    </Flex>
+  );
+};
+
+const getLastSeenValue = ({
+  isOnline,
+  lastSeenAt,
+}: {
+  isOnline: boolean | undefined;
```
Comment (reichert621) [inline assets/src/components/customers/CustomerDetailsSidebar.tsx:282]: i _think_ in typescript it's a bit more idiomatic to use `isOnline?: boolean` instead of `isOnline: boolean | undefined` (i'm pretty sure they're the same thing, but i might be wrong)

File: assets/src/components/customers/CustomerUtils.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import * as API from '../../api';
```
Comment (reichert621) [inline assets/src/components/customers/CustomerUtils.ts:1]: nit -- for the file name, we've been using the convention of calling util files like this `customers/support.ts` (just a convention we used at stripe which i've kept)

File: assets/src/components/customers/CustomersTable.tsx [hunk 9]
```diff
@@ -4,7 +4,7 @@ import dayjs from 'dayjs';
 import utc from 'dayjs/plugin/utc';
 import {Customer} from '../../types';
 import {Badge, Button, Table, Text, Tooltip} from '../common';
-import CustomerDetailsModal from './CustomerDetailsModal';
+import CustomerDetailsModal from '../customers/CustomerDetailsModal';
```
Comment (reichert621) [inline assets/src/components/customers/CustomersTable.tsx:7]: was this change necessary? 🤔

Comment (reichert621) [reply assets/src/components/customers/CustomersTable.tsx:7]: (on an unrelated note, it might be interesting to switch to using absolute(?) paths rather than relative paths)

File: lib/chat_api_web/views/customer_view.ex (hunk 4 above)
Comment (reichert621) [reply lib/chat_api_web/views/customer_view.ex:54]: seems reasonable!

**Review by reichert621 — CHANGES_REQUESTED** [review]
looks great @rhonsby! very excited about these design improvements 🔥 

just some nitpicky stuff, curious to hear your thoughts :)

File: assets/src/components/customers/CustomerDetailsCard.tsx [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import {Box} from 'theme-ui';
+import {SxStyleProp} from 'theme-ui';
```
[the remark this answers is not in the document]
Comment (rhonsby) [reply assets/src/components/customers/CustomerDetailsCard.tsx:3]: Yeah, absolutely! Good catch.

File: assets/src/components/customers/CustomerDetailsSidebar.tsx [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+        <CustomerDetailsSection title="Tags">
+          <SidebarCustomerTags customerId={customerId} />
+        </CustomerDetailsSection>
+      </Box>
+    </CustomerDetailsCard>
+  );
+};
+
+const CompanyDetailsSection = ({company}: {company: Company | undefined}) => {
+  let body;
```
[the remark this answers is not in the document]
Comment (rhonsby) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx:173]: Yeah, sure thing :)

File: assets/src/components/customers/CustomerDetailsSidebar.tsx (hunk 5 above)
Comment (rhonsby) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx:242]: It's been a hot minute so I forgot that that was a best practice 😅 It didn't seem to be a problem, but I'm happy to change it to lowercase. 

Thoughts on a helper method used to reduce duplication? 

```
const getIconForComponent = (Icon) => (<Icon style={{color: color.primary}}>);
```

File: assets/src/components/customers/CustomerDetailsSidebar.tsx (hunk 6 above)
Comment (rhonsby) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx:250]: I considered that as well. I had it set up this way to reduce code duplication, but I'm not actually a huge fan of code paths where different types render different results (harder to reason about). Let's err on the side of being explicit. I'll change it 👌

File: assets/src/components/customers/CustomerDetailsSidebar.tsx (hunk 7 above)
Comment (rhonsby) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx:282]: Yeah, you're right. In this case, it isn't a big deal so I can change it to shorthand it, but there's one case where it actually does make a difference.

```
const firstFn = ({a, b}: {a: string, b: string | undefined}) => {
    console.log({a, b});
};

const secondFn = ({a, b}: {a: string, b?: string}) => {
    console.log({a, b});
};
```

In `firstFn`, `b` is required, but can be a `string` or `undefined`. In `secondFn`, `b` isn't required, but if it's there, it can be either `string` or `undefined`. 

How often does it happen where a property is required, but can be `undefined`? I'm not sure, but it's good to know the distinction. Again, for this case, it doesn't matter 🙂

File: assets/src/components/customers/CustomerUtils.ts (hunk 8 above)
Comment (rhonsby) [reply assets/src/components/customers/CustomerUtils.ts:1]: Ooo ok. I dig it. Will update!

File: assets/src/components/customers/CustomersTable.tsx (hunk 9 above)
Comment (rhonsby) [reply assets/src/components/customers/CustomersTable.tsx:7]: No, I'm not sure why this changed. I must have moved some stuff around and accidentally made this change. I'll fix it. 

re: absolute paths - I don't have an opinion either way.

File: assets/src/components/customers/CustomerDetailsSidebar.tsx (hunk 5 above)
Comment (reichert621) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx]: down for that! :+1:

File: assets/src/components/customers/CustomerDetailsSidebar.tsx (hunk 7 above)
Comment (reichert621) [reply assets/src/components/customers/CustomerDetailsSidebar.tsx]: ahh good point, i'd say if a property can be undefined than it's nice to have it not be required :)

**Review by reichert621 — APPROVED** [review]
nice! 🎉
